File: leetcode/fb_balanced_split.py
import random

# https://leetcode.com/discuss/interview-question/718692/facebook-training-balanced-split
# Looks like https://leetcode.com/problems/partition-equal-subset-sum/ but DIFFERENT!
# Find if a subset (LEFT) sums to half the total, all LEFT < RIGHT
# The following (quickselect) solution works for this specification, see also cpp


def balancedSplitExists(arr):
    # Write your code here
    s = sum(arr)
    if s % 2 == 1:
        return False
    goal = s // 2
    while True:
        pivot = arr[random.randint(0, len(arr) - 1)]
        lHalf = []
        rHalf = []
        for v in arr:
            if v <= pivot:
                lHalf.append(v)
            else:
                rHalf.append(v)

        lSumPivot = sum(lHalf)

        # key idea: we can't split the elements equal to pivot
        if lSumPivot == goal:
            return True

        if lSumPivot < goal:
            arr = rHalf
            goal -= lSumPivot
            continue

        # if LEFT <= RIGHT is allowed, then we can split the pivot elements
        # so need to decide how to split them. Here our we are just checking if left
        # are all equal to pivot, which is not that efficient (see cpp solution)
        # because we don't reduce arr's size in this case
        if all(i == lHalf[0] for i in lHalf):
            return False  # there's a blob in the middle we can't separate

# These are the tests we use to determine if the solution is correct.
# You can add your own at the bottom.


def printString(string):
    print('[\"', string, '\"]', sep='', end='')


test_case_number = 1


def check(expected, output):
    global test_case_number
    result = False
    if expected == output:
        result = True
    rightTick = '\u2713'
    wrongTick = '\u2717'
    if result:
        print(rightTick, 'Test #', test_case_number, sep='')
    else:
        print(wrongTick, 'Test #', test_case_number,
              ': Expected ', sep='', end='')
        printString(expected)
        print(' Your output: ', end='')
        printString(output)
        print()
    test_case_number += 1


if __name__ == "__main__":
    arr_1 = [1, 3]
    expected_1 = False
    output_1 = balancedSplitExists(arr_1)
    check(expected_1, output_1)

    arr_2 = [4, 4, 4, 4]
    expected_2 = False
    output_2 = balancedSplitExists(arr_2)
    check(expected_2, output_2)

    arr_3 = [2, 1, 2, 5]
    expected_3 = True
    output_3 = balancedSplitExists(arr_3)
    check(expected_3, output_3)

    arr_4 = [1, 5, 7, 1]
    expected_4 = True
    output_4 = balancedSplitExists(arr_4)
    check(expected_4, output_4)

    arr_5 = [3, 6, 3, 4, 4]
    expected_5 = False
    output_5 = balancedSplitExists(arr_5)
    check(expected_5, output_5)

    arr_6 = [5, 7, 8]
    expected_6 = False
    output_6 = balancedSplitExists(arr_6)
    check(expected_6, output_6)

    # Add your own test cases here

File: leetcode/fb_balanced_split.cpp
#include <span>
#include <random>
#include <vector>
#include <iostream>

class Rand_int {
public:
    Rand_int(int low, int high)
        : dist(low, high) {}
    double operator()() { return dist(re); } // draw an int
    void seed(int s) { re.seed(s); } // choose new random engine seed
private:
    std::default_random_engine re;
    std::uniform_real_distribution<> dist;
};

// Experimental C++20 solution
class Solution {
public:
    bool balancedSplit(std::vector<int>&& nums) {
        int s = std::reduce(nums.begin(), nums.end());
        if (s % 2 == 1) {
            return false;
        }

        // we want to find a left partition that sums to goal
        int goal = s / 2; // we know s / 2 is actually an integer
        Rand_int rnd { 0, 1 };

        std::span<int> nums_span(nums);
        while (s > goal) {
            int idx = rnd() * nums_span.size(); // round down to int
            int v = nums_span[idx]; // our pivot
            
            // we want three groups: [0, less), [less, greater), [greater, max)
            auto greater = std::partition(nums_span.begin(), nums_span.end(), 
                            [v](int x){ return x <= v;});
            auto less = std::partition(nums_span.begin(), greater,
                            [v](int x){ return x < v;});

            int rSum = std::reduce(greater, nums_span.end());
            int lSumPivot = s - rSum;

            if (lSumPivot > goal) { // case 1: recurse on left
                nums_span = std::span<int>(nums_span.begin(), less);
                
                // pivot goes right, need to move some elements from left to right
                int pivotSum = (greater - less) * v;
                s = lSumPivot - pivotSum; 
            } else { // case 2: recurse on right
                nums_span = std::span<int>(greater, nums_span.end());
                
                // pivot goes left, need to move some elements from right to left
                s = rSum;
                goal -= lSumPivot; 
            }
        }

        return goal == s;
    }
};

// output: 1,0,1,0,0,1,0,0,1
int main() {
    Solution s;
    std::cout << s.balancedSplit({1,5,7,1}) << std::endl;
    std::cout << s.balancedSplit({12,7,6,7,6}) << std::endl;
    std::cout << s.balancedSplit({}) << std::endl;
    std::cout << s.balancedSplit({2}) << std::endl;
    std::cout << s.balancedSplit({20,2}) << std::endl;
    std::cout << s.balancedSplit({5,7,20,12,5,7,6,14,5,5,6}) << std::endl;
    std::cout << s.balancedSplit({5,7,20,12,5,7,6,7,14,5,5,6}) << std::endl;
    std::cout << s.balancedSplit({1,1,1,1,1,1,1,1,1,1,1,1}) << std::endl;
    std::cout << s.balancedSplit({0,0,0}) << std::endl;
    
    return 0;
}